File: typescript/binary-search/binary-search.ts
export function find(haystack: number[], needle: number, left: number=0, right: number=(haystack.length-1)): number | never {
  let mid = Math.floor((left + right) / 2)

  switch (true) {
    case haystack[mid] === needle:
      return mid
    case right - left === 0:
      break
    case haystack[mid] < needle:
      return find(haystack, needle, mid+1, right)
    case haystack[mid] > needle:
      return find(haystack, needle, left, mid-1)
  }
  
  throw new Error('Value not in array')
}

File: ocaml/binary-search/binary_search.ml
open Base

let find (xs : int array) (value : int) : (int, string) result =
  let rec go lo hi =
    if lo > hi then Error "value not in array"
    else 
      let mid = lo + (hi - lo) / 2 in
      let mid_val = xs.(mid) in
      if mid_val < value then
        go (mid + 1) hi
      else if mid_val > value then
        go lo (mid - 1)
      else
        Ok mid
  in
  if Array.is_empty xs then Error "value not in array"
  else go 0 (Array.length xs - 1)
